### PR TITLE
feat: use debain to build apisix dev

### PR DESCRIPTION
**Pull request by tzssangglass**

### CODE REVIEW

**Comment by tzssangglass** [comment]
this PR based on: https://github.com/api7/apisix-build-tools/pull/199

**Review by tokers — APPROVED** [review]

File: debain-dev/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    && apt-get -y install git \
+        curl \
+        unzip \
+        lua5.1 \
+        liblua5.1-0-dev \
+        libldap2-dev \
+        gawk \
+    && mv /usr/bin/awk /usr/bin/awk.bak \
+    && mv /usr/bin/gawk /usr/bin/awk \
+    && git config --global url.https://github.com/.insteadOf git://github.com/ \
```
Comment (spacewander) [inline debain-dev/Dockerfile:38]: Why do we need this configuration?

File: .github/workflows/apisix-docker-test.yaml [hunk 2]
```diff
@@ -27,7 +27,7 @@ jobs:
       - uses: actions/checkout@v2
 
       - name: Clone apisix for local test
-        if: ${{ matrix.platform == 'alpine-local' }}
+        if: ${{ matrix.platform == 'debain-local' }}
```
Comment (spacewander) [inline .github/workflows/apisix-docker-test.yaml:30]: It should be `debian`

File: debain-dev/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ARG ENABLE_PROXY=false
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN set -x \
+    && (test "${ENABLE_PROXY}" != "true" || /bin/sed -i 's,http://archive.ubuntu.com,http://mirrors.aliyun.com,g' /etc/apt/sources.list) \
+    && (test "${ENABLE_PROXY}" != "true" || /bin/sed -i 's,http://ports.ubuntu.com,http://mirrors.aliyun.com,g' /etc/apt/sources.list) \
+    && apt-get -y update --fix-missing \
+    && apt autoremove -y libldap-2.4-2 git git-man \
+    && apt-get -y install git \
```
Comment (spacewander) [inline debain-dev/Dockerfile:29]: It is strange to remove git and then install it back

File: debain-dev/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#
+
+FROM api7/apisix-base:dev AS production-stage
+
+ARG ENABLE_PROXY=false
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN set -x \
+    && (test "${ENABLE_PROXY}" != "true" || /bin/sed -i 's,http://archive.ubuntu.com,http://mirrors.aliyun.com,g' /etc/apt/sources.list) \
```
Comment (spacewander) [inline debain-dev/Dockerfile:25]: We should replace the source list of debian, not ubuntu

File: debain-dev/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    && apt-get -y update --fix-missing \
+    && apt autoremove -y libldap-2.4-2 git git-man \
+    && apt-get -y install git \
+        curl \
+        unzip \
+        lua5.1 \
+        liblua5.1-0-dev \
+        libldap2-dev \
+        gawk \
+    && mv /usr/bin/awk /usr/bin/awk.bak \
```
Comment (spacewander) [inline debain-dev/Dockerfile:36]: As awk is a virtual package, install gawk will automatically replace /usr/bin/awk

File: debain-dev/Dockerfile (hunk 4 above)
Comment (tzssangglass) [reply debain-dev/Dockerfile:25]: but I found `deb http://archive.ubuntu.com/ubuntu/ focal main restricted` in `/etc/apt/sources.list`, and use

`cat /etc/apt/sources.list | grep debain` to get nothing.

File: debain-dev/Dockerfile (hunk 3 above)
Comment (tzssangglass) [reply debain-dev/Dockerfile:29]: fixed

File: debain-dev/Dockerfile (hunk 5 above)
Comment (tzssangglass) [reply debain-dev/Dockerfile]: fixed

File: debain-dev/Dockerfile (hunk 1 above)
Comment (tzssangglass) [reply debain-dev/Dockerfile]: rm it

File: .github/workflows/apisix-docker-test.yaml (hunk 2 above)
Comment (tzssangglass) [reply .github/workflows/apisix-docker-test.yaml]: fixed all

File: debain-dev/Dockerfile (hunk 4 above)
Comment (tzssangglass) [reply debain-dev/Dockerfile]: fixed

File: debian-dev/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+FROM api7/apisix-base:dev AS production-stage
+
+ARG ENABLE_PROXY=false
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN set -x \
+    && (test "${ENABLE_PROXY}" != "true" || /bin/sed -i 's,http://deb.debian.org,http://mirrors.aliyun.com,g' /etc/apt/sources.list) \
+    && apt-get -y update --fix-missing \
+    && apt autoremove -y libldap-2.4-2 \
```
Comment (spacewander) [inline debian-dev/Dockerfile:27]: Why do we need to remove libldap-2.4-2 here?

Comment (tzssangglass) [reply debian-dev/Dockerfile]: delete it

File: debian-dev/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    && curl https://raw.githubusercontent.com/apache/apisix/master/utils/linux-install-luarocks.sh -sL | bash - \
+    && luarocks install https://github.com/apache/apisix/raw/master/rockspec/apisix-master-0.rockspec --tree=/usr/local/apisix/deps PCRE_DIR=/usr/local/openresty/pcre \
+    && cp -v /usr/local/apisix/deps/lib/luarocks/rocks-5.1/apisix/master-0/bin/apisix /usr/bin/ \
+    && mv /usr/local/apisix/deps/share/lua/5.1/apisix /usr/local/apisix \
+    # forward request and error logs to docker log collector
+    && ln -sf /dev/stdout /usr/local/apisix/logs/access.log \
+    && ln -sf /dev/stderr /usr/local/apisix/logs/error.log
+
+FROM turien/apisix-base:dev AS production-stage
+
```
Comment (soulbird) [inline debian-dev/Dockerfile:46]: api7/apisix-base:dev ?

File: debian-dev/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+FROM api7/apisix-base:dev AS build
+
+ARG ENABLE_PROXY=false
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN set -x \
+    && (test "${ENABLE_PROXY}" != "true" || /bin/sed -i 's,http://deb.debian.org,http://mirrors.aliyun.com,g' /etc/apt/sources.list) \
+    && apt-get -y update --fix-missing \
+    && apt-get install -y git \
```
Comment (tao12345666333) [inline debian-dev/Dockerfile:27]: Typically, it is recommended that these names be in alphabetical order

File: debian-dev/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        sudo \
+    && curl https://raw.githubusercontent.com/apache/apisix/master/utils/linux-install-luarocks.sh -sL | bash - \
+    && luarocks install https://github.com/apache/apisix/raw/master/rockspec/apisix-master-0.rockspec --tree=/usr/local/apisix/deps PCRE_DIR=/usr/local/openresty/pcre \
+    && cp -v /usr/local/apisix/deps/lib/luarocks/rocks-5.1/apisix/master-0/bin/apisix /usr/bin/ \
+    && mv /usr/local/apisix/deps/share/lua/5.1/apisix /usr/local/apisix \
+    # forward request and error logs to docker log collector
+    && ln -sf /dev/stdout /usr/local/apisix/logs/access.log \
+    && ln -sf /dev/stderr /usr/local/apisix/logs/error.log
+
+FROM api7/apisix-base:dev AS production-stage
```
Comment (tao12345666333) [inline debian-dev/Dockerfile:45]: `api7/apisix-base:dev` is built base on ubuntu, not debian, right?

Comment (tzssangglass) [reply debian-dev/Dockerfile:45]: yes, would be fixed by:https://github.com/api7/apisix-build-tools/pull/209/files, let's wait for more time.

File: debian-dev/Dockerfile (hunk 8 above)
Comment (tzssangglass) [reply debian-dev/Dockerfile]: fixed

**Review by spacewander — APPROVED** [review]

**Pull request merged**